Fix order of source folders in classpath.

rgladwell opened the pull request:
During classpath configuration the order of source code folders changes
causing problems with source code lookup, for example during debugging:

https://github.com/rgladwell/m2e-android/issues/issue/245

The order should always be:

src/main/java
src/main/resources
src/test/java
src/test/resources
Android Framework
Android Private Libraries
Android Dependencies
Maven Dependencies
Non-runtime Maven Dependencies

This change reorders the classpath as above. Includes tests.

WonderCsabo commented:
@rgladwell this PR totally lacks lots of [classpath entries](https://github.com/rgladwell/m2e-android/pull/248/files#diff-1c257d9207f2250ead41f0dd5133eecdR67), which should be ordered properly. Did you intentionally skip those?

rgladwell commented:
I find these code reviews go better if we try to politely discuss both the positives and negatives.

Depends what you mean by "lacks lots of": after your commit that refactored the classpath API it seems most of the ordering issues were resolved except for the `/gen`source folder, and the business code changes are mostly for this.

The tests only verify that the `src/main`, 'gen', 'src/test' and classpath containers (as a group) are in the correct order. It doesn't verify the `resources` folders. This is sufficient to test the change I've made, but perhaps not sufficient to prevent classpath order regressions in the future. I can add more conditions if you think this would be necessary.

WonderCsabo commented:
Ricardo, I meant to be polite. Did you find my tone  too hard?  I do not, an if it is, that due to translation, I guess.

I think it would be more safe to explicitly set the order here and do not rely on the other code which may order the entries by luck. 

Also please note the last change adds the Android entries in a work space configurer, so ordering should be after that.

rgladwell commented:
No worries, I appreciate English isn't your first language.

When you say "set the order" do you mean in the tests?

I'm not sure how much detail to set the order: down to the individual order of each classpath container, or just the order of the `src/*` folders including `src/main/java`?

Is there any reason to believe a `WorkspaceConfigurer` is called after a `RawClasspathConfigurer`?

WonderCsabo commented:
> When you say "set the order" do you mean in the tests?

Now, i would order all the entries in the configurer just like @kingargyle's old implementation did.

> Is there any reason to believe a WorkspaceConfigurer is called after a RawClasspathConfigurer?

I am not really sure, however that was the only way when i created 90b5c7c. When i did it in the `90b5c7c`, the Android containers were not available, yet.

rgladwell commented:
> Now, i would order all the entries in the configurer just like @kingargyle's old implementation did.

I think additional code to order the classpath would be redundant after your refactor of the classpath API. It appears only the `/gen` folder is unordered.

I can certainly strengthen the tests to prevent regressions though.

Does this also mean that the order of the classpath containers themselves is important? Or only that they appear after the source code folders?

> I am not really sure, however that was the only way when i created 90b5c7c.

I am uncertain myself although the tests seem to be passing.

WonderCsabo commented:
I'm with writing tests for now only to verify the order. Good point!

WonderCsabo commented:
The order of the containers is not important for compilation, but it is
better to see a fixed, meaningful order in the gui.

rgladwell commented:
> better to see a fixed, meaningful order in the gui.

I'm not sure how @kingargyle derived the order of the classpath containers.

rgladwell commented:
Fixed PR, I decided not to test the order of the classpath containers as I'm unsure as to what is the correct order at this time.

Also added some minor refactorings.

@WonderCsabo please let me know what you think.

WonderCsabo commented on me.gladwell.eclipse.m2e.android/src/main/java/me/gladwell/eclipse/m2e/android/configuration/classpath/ReorderClasspathConfigurer.java:
```diff
@@ -0,0 +1,50 @@
+            IJavaProject javaproject = JavaCore.create(eclipseProject.getProject());
+            IClasspathEntry[] tosort = javaproject.readRawClasspath();
+            IClasspathEntry[] sorted = sortClasspath(mavenProject, tosort);
+            javaproject.setRawClasspath(sorted, new NullProgressMonitor());
```
I am not sure this is the good approach. You could just modify the `IClasspathDescriptor` here. Then m2e will persist its contents.

rgladwell replied:
I agree. However, I'm trying to experiment with alternatives, but the m2e `IClasspathDescriptor` class doesn't allow you to update all its entries, and is basically not easy to order. Instead you have to remove all entries and then add them again in the right order which is also ugly and seems to drop the `/gen` folder for some reason.

Unfortunately, the Java sort algorithms also only seem to work on arrays, not sure if there is some other way we could order it.

Any ideas?

WonderCsabo replied:
> Unfortunately, the Java sort algorithms also only seem to work on arrays, not sure if there is some other way we could order it.

You can sort any `List` w/ `Collections.sort()`. Just use it on `descriptor.getEntryDescriptors()`.

rgladwell replied:
Good point: I did try this and the code is much simpler. However, I think we're going to have to call `javaproject.readRawClasspath` because the `IClasspathDescriptor` that is passed to this method is stale and lacks a `gen/` entry.

When the `gen/` is added it breaks the order of the classpath.

WonderCsabo replied:
Yeah, that is the problem i faced when i wrote the m2e 1.6 support, which i mentioned to you earlier.

rgladwell replied:
Cool, at least we've verified it twice. Maybe we should raise with upstream if we haven't already.

I'll revert my changes but tidy the code.

WonderCsabo replied:
This is not an upstream problem. This only can we fixed if m2e would run
lot of configuration cycles, to make sure other pulgins already contributed
to the classpath. I think m2e should not modify the ClasspathDescriptor in
while configuration cycle.

rgladwell replied:
I'm confused: did you mean "m2e should not modify the ClasspathDescriptor" or "m2e-android should not modify the ClasspathDescriptor"?

WonderCsabo replied:
I was talking about the upstream, m2e. So basically if it would update the
descriptor concurrently after the plugins change the classpath, everything
would break for sure. The current behavior us better.

WonderCsabo commented on me.gladwell.eclipse.m2e.android/src/main/java/me/gladwell/eclipse/m2e/android/configuration/SourceType.java:
```diff
@@ -0,0 +1,84 @@
+        }
+        return false;
+    }
+}
```
Cool solution for sorting. :+1:

rgladwell replied:
Thanks, should make it easier to update the order in future.

kingargyle commented:
Actually in some cases Order is necessary for compilation and also for some tooling to beable to look up the source code for a project.  Especially if an Eclipse Plugin is only checking for the first source folder when looking up the code, then it is critical that the src/main/java always shows up first.

Also, my experience when I was writing the original sorter, is that the particular configurator could run about 3 or 4 times and if you have other maven plugins also configuring and adding their entries, you could get stuff out of order again.  Especially if you are adding generated-sources and test-generated-sources to the list of entries.

rgladwell commented:
@kingargyle do we know if changing the order of the classpath containers would cause problems or is it OK leave this?

WonderCsabo commented:
@kingargyle 

> Especially if an Eclipse Plugin is only checking for the first source folder when looking up the code, then it is critical that the src/main/java always shows up first.

Indeed! That is already handled in this PR.

> Especially if you are adding generated-sources and test-generated-sources to the list of entries.

Good point, i also experienced that. It is more safe to order all the entries, not just the source folders.

rgladwell commented:
PR updated.

kingargyle commented:
The only order I think has to occur @rgladwell is that the Android SDK should be the first classpath container so that ADT/Andmore finds that in case it is order dependent (haven't really looked).  Some of the other is preference, but I'd like to have Maven Runtime, and then Non Maven Runtime in that order but not sure it really matters unless you have different libraries for Tests.

WonderCsabo commented:
@rgladwell i am afraid this PR is not correct. The problem is you read the raw classpath, then persist it in your configurer. However, after all raw classpath configurers ran, m2e-core will persist the raw classpath according to the `ClasspathDescriptor`. I think it will overwrite your changes. :(

rgladwell commented:
@WonderCsabo are you seeing problems on your local workspace?

WonderCsabo commented:
This is only a theory. I will try to verify this at home.

rgladwell commented:
Seems to be working locally for me.

Perhaps a more future-proof iteration of this would be to do the following:

1. Change the `AddGenFolderClasspathConfigurer` to be a `RawClasspathConfigurer` so it runs before the `ReorderClasspathConfigurer`.
2. Update `ReorderClasspathConfigurer` to directly order the `IClasspathDescriptor` as you suggest above.

This way we might be able to avoid directly writing the raw classpath and working with m2e.

I'll give it a try

WonderCsabo commented:
Try to add a new entry as the first entry in the last raw classpath
configurer to the descriptor.
On Jul 15, 2015 16:23, "Ricardo Gladwell" <notifications@github.com> wrote:

> Seems to be working locally for me.
>
> Perhaps a more future-proof iteration of this would be to do the following:
>
>    1. Change the AddGenFolderClasspathConfigurer to be a
>    RawClasspathConfigurer so it runs before the ReorderClasspathConfigurer
>    .
>    2. Update ReorderClasspathConfigurer to directly order the
>    IClasspathDescriptor as you suggest above.
>
> This way we might be able to avoid directly writing the raw classpath and
> working with m2e.
>
> I'll give it a try
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rgladwell/m2e-android/pull/364#issuecomment-121632266>
> .
>

rgladwell commented:
Sorry, I didn't understand that?

rgladwell commented:
I attempted my suggestion above but it didn't work: lots of test failures.

WonderCsabo commented:
Sorry, I quickly wrote down that in email. I will test it shortly after I
arrive home.

WonderCsabo commented:
@rgladwell sorry for my "theory", i think it is not true. This configurer is called when the classpath is resolved, so that is totally cool.

I also figured out why the `IClasspathDescriptor` cannot be used: `IJavaProjectConfigurator.configureClasspath` is used to configure the *Maven* classpath actually, so the descriptor only contains the Maven classpath entries.

WonderCsabo commented:
:( This worked for my real project, but now *target/generated-sources/annotations* is added as the first entry.

rgladwell commented:
@WonderCsabo should they appear in the same order as the `/gen` folder?
